### PR TITLE
web-scraping implementation

### PR DESCRIPTION
**Pull request by chinazor-allen**
### What does this PR do?
- Scrapes webpage and returns xpath or css selectors in xml or json format.

### CODE REVIEW

File: socket.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+name: simple-web-scraper
+description: Description of simple-web-scraper
```
Comment (adamwardecki) [inline socket.yml]: Update the description

File: README.md [hunk 2]
```diff
@@ -1 +1,52 @@
...
+### scraping
+
+Scrapes webpage and returns xpath or css selectors in xml or json format.
+
+#### Parameters
+
+| name | type | description | example | long_description
+| ---- | ---- | ----------- | ------- | ----------------
+| url | string | Webpage address to scrape from | https://funnycatsgallery.com/' | 
+| selectorType | string | Indicates the type of selector to use in scraing | xpath | Contents can be scraped using either `xpath` or `css` selector.
```
Comment (adamwardecki) [inline README.md]: s/scraing/scraping

File: socket.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+version: 0.0.1
+
+endpoints:
+  scraping:
+    description: Scrapes webpage and returns xpath or css selectors in xml or json format.
+    parameters:
+      url:
+        type: string
+        description: "Webpage address to scrape from"
+        example: "https://funnycatsgallery.com/'"
```
Comment (adamwardecki) [inline socket.yml]: There's an extra single quote after the url.

File: src/utility.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  if (selectorType === 'xpath') {
+    const xpathScraping = openScraping.parse(config, html);
+    return resolve(report(extract, xpathScraping));
+  } if (selectorType === 'css') {
+    try {
+      const $ = cheerio.load(html);
+
+      const configValue = Object.keys(config).map((value) => {
+        const decodedCss = [];
+        const isCssValue = $(config[value]).each((i, cssValue) => {
```
Comment (adamwardecki) [inline src/utility.js]: declaring isCssValue const is not needed here

File: src/utility.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+          decodedCss[i] = `${$(cssValue).text()}`;
```
Comment (adamwardecki) [inline src/utility.js]: $(cssValue).html()

File: src/utility.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        });
+
+        try {
+          if (Object.keys(config).length >= 1) {
+            config[value] = $(config[value]).text() === '' ?
+              `${config[value]} does not exist` :
+              decodedCss.toString();
```
Comment (adamwardecki) [inline src/utility.js]: Probably just `decodedCss`

File: socket.yml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      extract:
+        type: string
+        description: "Scraped Data readable format to return."
+        long_description: |
+          Data can be returned either in `json` or `xml` format.
+        example: "json"
+      config:
+        type: object
+        description: "Parameter for selecting the content you want to scrape."
+        example: |
```
Comment (adamwardecki) [inline socket.yml]: The example says that I should pass a `config` object along with rest of the params. My initial understanding was that the whole params object would look like this:

```js
{"url": "https://funnycatsgallery.com/", "selectorType": "css", "extract": "json", "config": {"image":"div.featured-thumbnail","title":"h1#logo"}}
```

but I it needs to be like so:

```js
{"url": "https://funnycatsgallery.com/", "selectorType": "css", "extract": "json", "image":"div.featured-thumbnail","title":"h1#logo"}
```

The first approach seems to be better.

File: src/scraping.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+  const {
+    url,
+    extract,
+    selectorType,
+    ...config
+  } = ctx.args;
+
+  try {
+    if (requestMethod === 'GET') {
```
Comment (adamwardecki) [inline src/scraping.js]: Why GET?

File: src/scraping.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import Syncano from 'syncano-server';
+import scraping from './utility';
+
+const rp = require('request-promise');
```
Comment (adamwardecki) [inline src/scraping.js]: `import rp from 'request-promise`

**Review by adamwardecki — CHANGES_REQUESTED** [review]
Problems that I see here:
1. The way of passing config object is not in line with the docs
2. The scraping (with css at least) returns an extracted text from the selected element. What would be more useful is an html string (or an array of html strings).
3. I noticed that when trying to pass a css class as a scraping parameter (i.e. "image":"div.featured-thumbnail"), I'm getting an error. It should be possible to do this and receive an array of elements that match this css

File: src/scraping.js (hunk 8 above)
Comment (chinazor-allen) [reply src/scraping.js]: A `POST` would be preferred as the user would send information to the server.

File: README.md (hunk 2 above)
Comment (chinazor-allen) [reply README.md]: done

File: socket.yml (hunk 1 above)
Comment (chinazor-allen) [reply socket.yml]: done

File: socket.yml (hunk 3 above)
Comment (chinazor-allen) [reply socket.yml]: done

File: src/scraping.js (hunk 8 above)
Comment (chinazor-allen) [reply src/scraping.js]: done

File: src/scraping.js (hunk 9 above)
Comment (chinazor-allen) [reply src/scraping.js]: done

File: src/utility.js (hunk 4 above)
Comment (chinazor-allen) [reply src/utility.js]: fixed

File: src/utility.js (hunk 5 above)
Comment (chinazor-allen) [reply src/utility.js]: done

File: src/utility.js (hunk 6 above)
Comment (chinazor-allen) [reply src/utility.js]: fixed

**Review by adamwardecki — APPROVED** [review]

**Review by adamwardecki — CHANGES_REQUESTED** [review]
The updated socket should also use the @syncano/cli package

**Review by adamwardecki — APPROVED** [review]

**Pull request merged**